File: clients/python/games/hex_game.py
from games.base import BaseGame
from framework.i18n import t

SIZE = 11


class Hex(BaseGame):
    min_players = max_players = 2

    def __init__(self):
        self.board = [[None] * SIZE for _ in range(SIZE)]
        self.players = []
        self._turn_idx = 0
        self._over = False
        self._winner = None

    def start(self, players):
        self.players = players  # [0]=black(top-bottom), [1]=white(left-right)

    def current_turn(self):
        return self.players[self._turn_idx] if self.players else None

    def _neighbors(self, r, c):
        return [(r + dr, c + dc)
                for dr, dc in [(-1, 0), (1, 0), (0, -1), (0, 1), (-1, 1), (1, -1)]
                if 0 <= r + dr < SIZE and 0 <= c + dc < SIZE]

    def _check_win(self, pid):
        if not self.players:
            return False
        if pid == self.players[0]:
            starts = [(0, c) for c in range(SIZE) if self.board[0][c] == pid]
            goal = lambda r, c: r == SIZE - 1
        else:
            starts = [(r, 0) for r in range(SIZE) if self.board[r][0] == pid]
            goal = lambda r, c: c == SIZE - 1
        visited = set()
        stack = list(starts)
        while stack:
            pos = stack.pop()
            if pos in visited:
                continue
            visited.add(pos)
            if goal(*pos):
                return True
            for nb in self._neighbors(*pos):
                if self.board[nb[0]][nb[1]] == pid and nb not in visited:
                    stack.append(nb)
        return False

    def validate_move(self, player_id, move_data):
        if self._over or player_id != self.current_turn():
            return False
        r, c = move_data.get('row'), move_data.get('col')
        return (r is not None and c is not None and
                0 <= r < SIZE and 0 <= c < SIZE and
                self.board[r][c] is None)

    def apply_move(self, player_id, move_data):
        r, c = move_data['row'], move_data['col']
        self.board[r][c] = player_id
        if self._check_win(player_id):
            self._over = True
            self._winner = player_id
        else:
            self._turn_idx = 1 - self._turn_idx

    def is_over(self):
        return self._over, self._winner

    def render(self, perspective=None):
        p0 = self.players[0] if self.players else 'black'
        p1 = self.players[1] if len(self.players) > 1 else 'white'
        syms = {None: '.', p0: 'B', p1: 'W'}
        lines = [t('hex.title')]
        for i, row in enumerate(self.board):
            lines.append(' ' * i + ' '.join(syms.get(c, '.') for c in row))
        return '\n'.join(lines)

    def get_state(self, perspective=None):
        return {'board': self.board, 'turn': self.current_turn(),
                'players': self.players}

File: clients/node/src/games/hex_game.js
'use strict';
const { BaseGame } = require('./base');
const { t } = require('../framework/i18n');

const SIZE = 11;

class Hex extends BaseGame {
  constructor() {
    super();
    this.board = Array.from({ length: SIZE }, () => Array(SIZE).fill(null));
    this.players = [];
    this._turnIdx = 0;
    this._over = false;
    this._winner = null;
  }

  start(players) {
    this.players = players;
    this._turnIdx = 0;
  }

  currentTurn() {
    return this.players[this._turnIdx] || null;
  }

  _neighbors(r, c) {
    return [[-1,0],[1,0],[0,-1],[0,1],[-1,1],[1,-1]]
      .map(([dr, dc]) => [r+dr, c+dc])
      .filter(([nr, nc]) => nr >= 0 && nr < SIZE && nc >= 0 && nc < SIZE);
  }

  _checkWin(pid) {
    if (!this.players.length) return false;
    const isP0 = pid === this.players[0];
    const starts = isP0
      ? Array.from({ length: SIZE }, (_, c) => [0, c]).filter(([, c]) => this.board[0][c] === pid)
      : Array.from({ length: SIZE }, (_, r) => [r, 0]).filter(([r]) => this.board[r][0] === pid);
    const goal = isP0 ? ([r]) => r === SIZE - 1 : ([, c]) => c === SIZE - 1;
    const visited = new Set();
    const stack = [...starts];
    while (stack.length) {
      const pos = stack.pop();
      const key = `${pos[0]},${pos[1]}`;
      if (visited.has(key)) continue;
      visited.add(key);
      if (goal(pos)) return true;
      for (const nb of this._neighbors(pos[0], pos[1])) {
        if (this.board[nb[0]][nb[1]] === pid && !visited.has(`${nb[0]},${nb[1]}`)) {
          stack.push(nb);
        }
      }
    }
    return false;
  }

  validateMove(playerId, moveData) {
    if (this._over || playerId !== this.currentTurn()) return false;
    const { row: r, col: c } = moveData;
    return r != null && c != null && r >= 0 && r < SIZE && c >= 0 && c < SIZE && this.board[r][c] === null;
  }

  applyMove(playerId, moveData) {
    const { row: r, col: c } = moveData;
    this.board[r][c] = playerId;
    if (this._checkWin(playerId)) {
      this._over = true;
      this._winner = playerId;
    } else {
      this._turnIdx = 1 - this._turnIdx;
    }
  }

  isOver() {
    return [this._over, this._winner];
  }

  render(perspective) {
    const [p0, p1] = [this.players[0] || 'black', this.players[1] || 'white'];
    const syms = { [p0]: 'B', [p1]: 'W', null: '.' };
    const lines = [t('hex.title')];
    for (let i = 0; i < SIZE; i++) {
      lines.push(' '.repeat(i) + this.board[i].map(c => syms[c] || '.').join(' '));
    }
    return lines.join('\n');
  }

  getState(perspective) {
    return { board: this.board, turn: this.currentTurn(), players: this.players };
  }

  loadState(data) {
    if (!data) return;
    if (data.players) this.players = data.players;
    if (data.board) this.board = data.board.map(row => [...row]);
    if (data.turn != null) {
      const idx = this.players.indexOf(data.turn);
      this._turnIdx = idx >= 0 ? idx : 0;
    }
  }
}

module.exports = { Hex };
